### ISSUE TITLE
Add guidance on pinning GitHub Actions and container images

### ISSUE THREAD

**Issue by pellared** [open-telemetry/sig-security#69]
## Why

A bad actor can force push a tag so that GitHub Action to do some malicious actions.

A bad actor can push a malicious container image under the same name.

## What

We should use digest pinning to mitigate the possibility of using a malicious GitHub Actions and container images. 
It should also make the build more reproducible.
Some references: 
- https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions
- https://candrews.integralblue.com/2023/09/always-use-docker-image-digests/

When using GitHub Actions we can add a comment with a after the digest (e.g. `actions/checkout@01aecccf739ca6ff86c0539fbc67a7a5007bbc81 # v2.1.0`).
Both Renovate and Dependabot can bump both the digest and the tag in the comment:
- https://github.com/dependabot/dependabot-core/issues/4691#issuecomment-1296532496
- https://docs.renovatebot.com/modules/manager/github-actions/

When using container images we can add the digest at the end (e.g. `node:14.15.1@sha256:d938c1761e3afbae9242848ffbb95b9cc1cb0a24d889f8bd955204d347a7266e`).
Both Renovate and Dependabot can bump the image name and the digest as well:
- https://github.com/dependabot/dependabot-core/pull/6150
- https://docs.renovatebot.com/docker/

**Comment by yurishkuro** [comment]
> Sounds like an automation issue? Renovate has a "noise reduction" feature, so that it would group related dependencies (like "all github actions") into one single PR

We are already using the grouping of dependencies in Jaeger, but it's not enough - if a project depends on a dozen GH actions there's a high chance at least one will have a new patch every week, which will create a new PR that provides exactly 0 value to me. So another thing I do is restrict certain groups to minor version bumps only & ignore patches.

**Comment by pellared** [comment]
> So another thing I do is restrict certain groups to minor version bumps only & ignore patches.

Patches use to have bugfixes (including security fixes). Not bumping them may not appropriate from security perspective.

**Comment by Kielek** [comment]
After introducing support for [digest in testcoinares](https://github.com/testcontainers/testcontainers-dotnet/releases/tag/4.0.0),
we have pinned both GH actions and docker images to SHA in OTel .NET Auto repo. No issues so far.

Ref:
 - GH action https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3505
 - Docker: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3514

**Comment by codeboten** [comment]
In light of https://semgrep.dev/blog/2025/popular-github-action-tj-actionschanged-files-is-compromised/, i propose we should revisit pinning versions for github actions (and containers) to hashses

**Comment by yurishkuro** [comment]
CNCF already recommends using `step-security/harden-runner` in the workflows. It provides reports when actions are not pinned by hashes

**Comment by codeboten** [comment]
@yurishkuro can you link where the recommendation is documented?

**Comment by yurishkuro** [comment]
ossf score card has Pinned Dependencies section, which links to https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#pinned-dependencies. There step-security check is recommended which will actually create PRs to pin all actions to hashes (after that dependency bots will keep them pinned during upgrades).

**Comment by mowies** [comment]
The same can also be achieved with Renovate using one of their presets: https://docs.renovatebot.com/modules/manager/github-actions/#digest-pinning-and-updating

**Comment by pellared** [comment]
@yurishkuro 

> CNCF already recommends using `step-security/harden-runner` in the workflows. It provides reports when actions are not pinned by hashes

Are you sure? Where is the information that `step-security/harden-runner` has such feature? I cannot find such infomration

> ossf score card has Pinned Dependencies section, which links to https://github.com/ossf/scorecard/blob/ea7e27ed41b76ab879c862fa0ca4cc9c61764ee4/docs/checks.md#pinned-dependencies. There step-security check is recommended which will actually create PRs to pin all actions to hashes (after that dependency bots will keep them pinned during upgrades).

I think you mean https://app.stepsecurity.io/ which is a separate tool than the one mentioned in CNCF docs and is also distinct from step-security/harden-runner` action.

**Comment by pellared** [comment]
> The same can also be achieved with Renovate using one of their presets: https://docs.renovatebot.com/modules/manager/github-actions/#digest-pinning-and-updating

Renovate recommends to use `"extends": ["config:best-practices"]` which already contains this preset 😉 

Reference: https://docs.renovatebot.com/upgrade-best-practices/#general-recommendations

**Comment by trask** [comment]
> In light of https://semgrep.dev/blog/2025/popular-github-action-tj-actionschanged-files-is-compromised/, i propose we should revisit pinning versions for github actions (and containers) to hashses

I agree with pinning versions (we implemented this in all of the Java repos recently), but one nagging question in my mind is how does this help us avoid picking up a compromised release, since we generally blindly hit merge on version bumps as long as CI is passing? Or is pinning versions just so that at least we have an accounting of exactly when we were running a compromised version in case a compromise is uncovered later?

**Comment by austinlparker** [comment]
I mean, it's a tradeoff? You could probably talk me into the idea that pinning versions is a better tradeoff as long as we're using good actions hygiene. In the case of changed-files, the impact was very low because we didn't leak unneeded secrets into the action. Leaking a generated GH_TOKEN that gets expired after the workflow runs strikes me as very low-impact... not to say that future attacks couldn't leverage the same process in order to do _other_ nasty stuff, though.

**Comment by pellared** [comment]
> Or is pinning versions just so that at least we have an accounting of exactly when we were running a compromised version in case a compromise is uncovered later?

This is very important for the incident response process.

> since we generally blindly hit merge on version bumps as long as CI is passing

This is not always the case. I was sometimes creating issues when some release was suspicious to me (e.g. missing GitHub Release or git tag was not signed).

**Comment by mowies** [comment]
> I agree with pinning versions (we implemented this in all of the Java repos recently), but one nagging question in my mind is how does this help us avoid picking up a compromised release, since we generally blindly hit merge on version bumps as long as CI is passing? Or is pinning versions just so that at least we have an accounting of exactly when we were running a compromised version in case a compromise is uncovered later?

@trask I think it's a step in the right direction. That way, a compromised GH action will not be used directly WITHOUT somebody actually merging it in. The issue with tj-actions/changed-files was that the attacker changed all tags to point to the malicious commit, which means that it ran that commit by default without manual intervention if the usage was not pinned to a different commit hash. That attack vector could be prevented.

The human element in merging a malicious version into the code manually can't be prevented of course 🤷🏼
